File: c/aoc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <ctype.h>
#include "aoc.h"

clock_t start, end;
double cpu_time_used;

int part = 0;
long long answer_a = 0;
long long answer_b = 0;

char Reversed[] =      "\033[7m";
char DimBackground[] = "\033[5m";
char Reset[] =         "\033[0m";

char ClearScreen[] =   "\033[2J";
char ClearLine[] =     "\033[2K";

char HideCursor[] =    "\033[?25l";
char ShowCursor[] =    "\033[?25h";

char BrightBlack[] =   "\033[30;1m";
char BrightRed[] =     "\033[31;1m";
char BrightGreen[] =   "\033[32;1m";
char BrightYellow[] =  "\033[33;1m";
char BrightBlue[] =    "\033[34;1m";
char BrightMagenta[] = "\033[35;1m";
char BrightCyan[] =    "\033[36;1m";
char BrightWhite[] =   "\033[37;1m";

void start_day(int day)
{
	printf("%s %sDay %s%d %s\n\n", DimBackground, BrightWhite, BrightYellow, day, Reset);
}

void start_part_a()
{
	printf("%sPart %sA%s\n", BrightWhite, BrightCyan, Reset);
	part = 1;
    start = clock();
}

void start_part_b()
{
	printf("%sPart %sB%s\n", BrightWhite, BrightCyan, Reset);
	part = 2;
    start = clock();
}

long long get_answer_a()
{
	return answer_a;
}

long long get_answer_b()
{
	return answer_b;
}

void show_answer(long long answer)
{
    end = clock();

	if (part == 1)
	{
		answer_a = answer;
		printf("Answer: %s%lld%s", BrightGreen, answer_a, Reset);
	}
	else if (part == 2)
	{
		answer_b = answer;
		printf("Answer: %s%lld%s", BrightGreen, answer_b, Reset);
	}

    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
	printf(" | Took %s%f%s seconds\n\n", BrightMagenta, cpu_time_used, Reset);
}

void free_input(char** lines, int line_count)
{
	for (int i = 0; i < line_count; i++)
	{
		if (lines[i] != NULL)
		{
			free(lines[i]);
		}
	}

	free(lines);
}

char** read_input(int day, char* path, int max_lines, int max_line_length, int* line_count)
{
	char filename[100];
	sprintf(filename, "%sinput-day%d.txt", path, day);
	FILE *file = fopen(filename, "r");
	if (file == NULL)
	{
		fprintf(stderr, "Could not open file %s\n", filename);
		perror("fopen()");
		return NULL;
	}

	char** lines = (char**)malloc(max_lines * sizeof(char*));
	for (int i = 0; i < max_lines; i++)
	{
		lines[i] = NULL;
	}

	size_t len = max_line_length;
	char* line = (char* )malloc(len);
	ssize_t nread;
	while ((nread = getline(&line, &len, file)) != -1)
	{
		// printf("Retrieved line of length %zu:\n", nread);
		// fwrite(line, nread, 1, stdout);
		char* thisline = (char*)malloc(nread * sizeof(char));
		lines[(*line_count)] = thisline;
		strcpy(thisline, line);
		(*line_count)++;
	}

	free(line);
	fclose(file);
	return lines;
}

int* convert_input_to_numbers(char **lines, int line_count)
{
	int* numbers = (int*)malloc(line_count * sizeof(int));
	for (int ix = 0; ix < line_count; ix++)
	{
		numbers[ix] = atoi(lines[ix]);
	}

	return numbers;
}

char *ltrim(char *s)
{
    while (isspace(*s)) s++;
    return s;
}

char *rtrim(char *s)
{
    char* back = s + strlen(s);
    while (isspace(*--back));
    *(back+1) = '\0';
    return s;
}

char *trim(char *s)
{
    return rtrim(ltrim(s));
}

File: python/aoc.py
import time
import os
from sys import platform, stdout

class Ansi():
    Reversed =      u"\u001b[7m"
    Underline =     u"\u001b[4m"
    SlowBlink =     u"\u001b[5m"
    Reset =         u"\u001b[0m"

    ClearScreen =   u"\u001b[2J"
    ClearLine =     u"\u001b[2K"

    StoreCursor =   u"\u001b[s"
    RestoreCursor = u"\u001b[u"

    HideCursor =    u"\u001b[?25l"
    ShowCursor =    u"\u001b[?25h"

    Black =         u"\u001b[30m"
    Red =           u"\u001b[31m"
    Green =         u"\u001b[32m"
    Yellow =        u"\u001b[33m"
    Blue =          u"\u001b[34m"
    Magenta =       u"\u001b[35m"
    Cyan =          u"\u001b[36m"
    White =         u"\u001b[37m"

    BrightBlack =   u"\u001b[30;1m"
    BrightRed =     u"\u001b[31;1m"
    BrightGreen =   u"\u001b[32;1m"
    BrightYellow =  u"\u001b[33;1m"
    BrightBlue =    u"\u001b[34;1m"
    BrightMagenta = u"\u001b[35;1m"
    BrightCyan =    u"\u001b[36;1m"
    BrightWhite =   u"\u001b[37;1m"

    BlackBackground =   u"\u001b[40m"
    RedBackground =     u"\u001b[41m"
    GreenBackground =   u"\u001b[42m"
    YellowBackground =  u"\u001b[43m"
    BlueBackground =    u"\u001b[44m"
    MagentaBackground = u"\u001b[45m"
    CyanBackground =    u"\u001b[46m"
    WhiteBackground =   u"\u001b[47m"

class Aoc():

    def __init__(self):
        self._day = 0
        self._start = 0
        self._end = 0
        self._part = 0
        self.inputdata = []
        self.AnswerA = None
        self.AnswerB = None

    def StartDay(self, day:int, title:str = "AOC") -> None:
        self._day = day
        self.title = title
        print(f"{Ansi.BlueBackground} {Ansi.White}Day {Ansi.BrightMagenta}{self._day}{Ansi.White} - {Ansi.Red}{self.title} {Ansi.Reset}")

    def ReadInput(self):
        self.inputdata.clear()
        if os.path.isfile(self.input_filename()) == False:
            print(f"{Ansi.BrightRed}File {self.input_filename()} not found{Ansi.Reset}")
            return

        file = open(self.input_filename(), "r")
        for line in file:
            self.inputdata.append(line.rstrip())
        file.close()

    def input_filename(self):
        return f"../input/input-day{self._day}.txt"

    def StartPartA(self):
        print("")
        print(f"{Ansi.Underline}{Ansi.BrightWhite}Part {Ansi.BrightCyan}A{Ansi.Reset}")
        self._start = time.perf_counter()
        self._part = 1

    def StartPartB(self):
        print("")
        print(f"{Ansi.Underline}{Ansi.BrightWhite}Part {Ansi.BrightCyan}B{Ansi.Reset}")
        self._start = time.perf_counter()
        self._part = 2

    def ShowAnswer(self, result):
        if self._part == 1:
            self.AnswerA = result
        else:
            self.AnswerB = result

        self._end = time.perf_counter()
        ellapsed = self._end - self._start
        print(f"Answer: {Ansi.BrightGreen}{result}{Ansi.Reset} | Took {Ansi.BrightMagenta}{ellapsed:.5f}{Ansi.Reset} seconds")
        print("")

    def GetAnswerA(self):
        return self.AnswerA

    def GetAnswerB(self):
        return self.AnswerB

    def MoveCursor(self, column: int, line: int) -> str:
        return u"\u001b[%d;%dH" % (line, column)

    def MoveCursorUp(self, lines: int) -> str:
        return u"\u001b[%dA" % lines

    def Flush(self):
        stdout.flush()

    def Assert(self, answer, goal):
        try:
            assert answer == goal
            print(f"{Ansi.GreenBackground}{Ansi.Black}OK{Ansi.Reset}")
            print("")
        except AssertionError as e:
            print(f"{Ansi.RedBackground}{Ansi.BrightWhite}Test Failed{Ansi.Reset} {Ansi.BrightRed}{answer}{Ansi.Reset} is not equal to {Ansi.BrightGreen}{goal}{Ansi.Reset}")
            print("")
        else:
            pass
        finally:
            pass

if platform == "win32":
    import msvcrt, ctypes
    kernel32 = ctypes.windll.kernel32
    kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)

## Main

if __name__ == "__main__":
    aoc = Aoc()
    aoc.StartPartA()
    time.sleep(1)
    aoc.ShowAnswer(42)
    aoc.StartPartB()
    time.sleep(0.4)
    aoc.ShowAnswer(80085)
